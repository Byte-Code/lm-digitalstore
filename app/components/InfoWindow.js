import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import BlockIcon from 'material-ui/svg-icons/navigation/close';
import glamorous from 'glamorous';

import { getStockLabel } from '../utils/utils';

export default class InfoWindow extends React.Component {
  static propTypes = {
    selectedStoreInfo: ImmutablePropTypes.map.isRequired,
    handleClick: PropTypes.func.isRequired
  };

  static getOffsets(element) {
    return {
      top: element.offsetTop,
      left: element.offsetLeft
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      top: -230,
      borderBottom: '0px',
      borderTop: '14px solid #fff',
      arrowTop: 'initial'
    };
    this.computePosition = this.computePosition.bind(this);
  }

  componentDidMount() {
    setTimeout(this.computePosition, 10);
  }

  computePosition() {
    const pin = document.getElementById('pin');
    const map = document.getElementById('map');
    const pinOffsets = {
      top: pin.parentElement.offsetTop,
      left: pin.parentElement.offsetLeft
    };
    const mapOffsets = {
      top: map.offsetTop,
      left: map.offsetLeft
    };

    if (pinOffsets.top < mapOffsets.top) {
      this.setState({
        top: 50,
        borderBottom: '14px solid white',
        borderTop: '0px',
        arrowTop: '-16px'
      });
    }

    if (pinOffsets.left > mapOffsets.left) {
      this.setState({ left: 100 });
    }
  }

  render() {
    const { selectedStoreInfo, handleClick } = this.props;
    const { top, borderBottom, borderTop, arrowTop, left } = this.state;
    const name = selectedStoreInfo.get('name');
    const availability = selectedStoreInfo.get('storeStock');
    const stockStatus = selectedStoreInfo.get('stockStatus');
    const isAvailable = availability > 0;
    const label = getStockLabel(availability, stockStatus);
    const street = selectedStoreInfo.getIn(['address', 'street']);
    const streetNumber = selectedStoreInfo.getIn(['address', 'streetNumber']) || '';
    const zip = selectedStoreInfo.getIn(['address', 'zipCode']);
    const city = selectedStoreInfo.getIn(['address', 'city']);
    const province = selectedStoreInfo.getIn(['address', 'state']);
    const arrowStyle = { arrowTop, borderTop, borderBottom, left };

    return (
      <Wrapper top={top} left={left} id="pin">
        <BlockIcon style={iconStyle} fill="#333333" onTouchTap={handleClick} />
        <div>
          <StoreName>{name}</StoreName>
          <Address>
            <p>{`${street} ${streetNumber}`}</p>
            <p>{`${city} - (${province})`}</p>
            <p>{zip}</p>
          </Address>
        </div>
        <Divider />
        <Stock isAvailable={isAvailable}>
          {label}
        </Stock>
        <Arrow {...arrowStyle} />
      </Wrapper>
    );
  }

}

const Wrapper = glamorous.div(({ top, left = 0 }) => ({
  width: '300px',
  backgroundColor: '#fff',
  boxShadow: '0 0 8px 0 #e4e4e4',
  position: 'absolute',
  top: `${top}px`,
  left: `-${left}px`,
  padding: '20px',
  zIndex: 1
}));

const Divider = glamorous.div({
  border: '1px dashed #333333',
  width: '100%'
});

const Stock = glamorous.div(({ isAvailable }) => ({
  marginTop: '15px',
  color: isAvailable ? '#67cb33' : '#333333',
  fontSize: '14px',
  fontFamily: 'LeroyMerlinSans Bold'
}));

const StoreName = glamorous.p({
  fontSize: '20px',
  color: '#262626',
  marginBottom: '10px',
  fontFamily: 'LeroyMerlinSans Bold'
});

const Arrow = glamorous.div(({ arrowTop, borderTop, borderBottom, left = 14 }) => ({
  width: 0,
  height: 0,
  borderLeft: '14px solid transparent',
  borderRight: '14px solid transparent',
  borderBottom,
  borderTop,
  position: 'absolute',
  left: `${left}px`,
  filter: 'drop-shadow(0 2px 0 #e4e4e4)',
  top: arrowTop,
  bottom: '-14px'
}));

const Address = glamorous.div({
  marginBottom: '20px',
  '&>p': {
    fontSize: '15px',
    color: '#333333'
  }
});

const iconStyle = {
  width: 24,
  height: 24,
  position: 'absolute',
  top: 5,
  right: 5,
  cursor: 'pointer'
};
