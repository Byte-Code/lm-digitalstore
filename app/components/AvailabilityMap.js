import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GoogleMapReact from 'google-map-react';
// import { meters2ScreenPixels } from 'google-map-react/utils';
import ImmutablePropTypes from 'react-immutable-proptypes';
import Slider from 'material-ui/Slider';
import { List, Map } from 'immutable';
import glamorous, { Div } from 'glamorous';

import Marker from './Marker';
import InfoWindow from './InfoWindow';
import NearbyStores from './NearbyStores';

export default class AvailabilityMap extends Component {
  static propTypes = {
    productName: PropTypes.string.isRequired,
    productCode: PropTypes.string.isRequired,
    selectedStore: PropTypes.string.isRequired,
    allNearbyStores: ImmutablePropTypes.list,
    homeStore: ImmutablePropTypes.map.isRequired,
    selectedStoreInfo: ImmutablePropTypes.map,
    nearbyStoresWithProductInStock: ImmutablePropTypes.list.isRequired,
    selectStore: PropTypes.func.isRequired,
    closeInfoWindow: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
    handleSlide: PropTypes.func.isRequired,
    radius: PropTypes.number.isRequired,
    zoom: PropTypes.number.isRequired,
    infoWindowOpen: PropTypes.bool.isRequired,
    trackStoreAvailabilityEvent: PropTypes.func.isRequired,
    stocks: ImmutablePropTypes.map
  };

  static defaultProps = {
    allNearbyStores: List(),
    selectedStoreInfo: Map(),
    stocks: Map()
  };

  constructor(props) {
    super(props);
    this.updateNearByList = false;
    this.state = {
      minRadius: 2,
      maxRadius: 50,
      sliderWidth: 960
    };
    this.nearByClick = this.nearByClick.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.selectedStoreInfo.equals(this.props.selectedStoreInfo)) {
      const storeName = nextProps.selectedStoreInfo.get('name');
      const storeCode = nextProps.selectedStoreInfo.get('code');
      const productCode = this.props.productCode;
      const storeStock = this.props.stocks.getIn([storeCode, productCode]);
      this.props.trackStoreAvailabilityEvent({ storeName, storeStock });
    }
  }

  getLabelPosition = radius => {
    const { sliderWidth, maxRadius, minRadius } = this.state;
    return (radius - minRadius) * sliderWidth / (maxRadius - minRadius); // eslint-disable-line
  };

  nearByClick(event) {
    this.props.selectStore(event);
    this.updateNearByList = true;
  }

  renderMarkers() {
    const { allNearbyStores, homeStore, selectStore, closeInfoWindow,
      stocks, productCode } = this.props;
    return allNearbyStores.map(s => {
      const code = s.get('code');
      const lat = s.get('latitude');
      const isAvailable = stocks.getIn([code, productCode]) > 0;
      const lng = s.get('longitude');
      const isCurrentStore = homeStore.get('code') === code;

      return (
        <Marker
          lat={lat}
          lng={lng}
          key={code}
          code={code}
          isCurrentStore={isCurrentStore}
          handleClick={() => {
            selectStore(code);
            closeInfoWindow();
            this.updateNearByList = false;
          }}
          isAvailable={isAvailable}
        />
      );
    });
  }

  renderInfoWindow() {
    const { selectedStoreInfo, selectedStore, infoWindowOpen,
      closeInfoWindow, stocks, productCode } = this.props;

    if (!infoWindowOpen || !selectedStore) {
      return null;
    }

    const lat = selectedStoreInfo.get('latitude');
    const lng = selectedStoreInfo.get('longitude');
    const stockValue = stocks.getIn([selectedStoreInfo.get('code'), productCode]);

    return (
      <InfoWindow
        lat={lat}
        lng={lng}
        handleClick={closeInfoWindow}
        selectedStoreInfo={selectedStoreInfo}
        storesStock={stockValue}
      />
    );
  }

  render() {
    const {
      productName,
      productCode,
      homeStore,
      nearbyStoresWithProductInStock,
      radius,
      zoom,
      selectedStore,
      handleChange,
      handleSlide,
      } = this.props;
    const { minRadius, maxRadius } = this.state;
    const homeStoreName = homeStore.get('name');
    const lat = homeStore.getIn(['gpsInformation', 'x']);
    const lng = homeStore.getIn(['gpsInformation', 'y']);
    const center = { lat, lng };
    const labelPosition = this.getLabelPosition(radius);

    return (
      <Wrapper>
        <Title>Verifica Disponibilit√†</Title>
        <ProductInfo>{`${productName} - REF. ${productCode}`}</ProductInfo>
        <Div padding="0 20px 40px" position="relative">
          <SliderTitle >{`Seleziona il raggio di distanza dal negozio di ${homeStoreName}`}</SliderTitle>
          <Slider
            min={minRadius}
            max={maxRadius}
            value={radius}
            onChange={handleSlide}
            color="#67cb33"
            sliderStyle={sliderStyle}
            step={10}
          />
          <Radius left={labelPosition}>{`${radius} km`}</Radius>
        </Div>
        <MapWrapper id="map">
          <GoogleMapReact
            center={center}
            fullscreenControl={false}
            options={mapOptions}
            onChange={handleChange}
            zoom={zoom}
          >
            {this.renderMarkers()}
            {this.renderInfoWindow()}
            <Circle lat={lat} lng={lng} width={930} height={930} />
          </GoogleMapReact>
          <ControlsOverlay />
        </MapWrapper>
        <NearbyStores
          nearbyStores={nearbyStoresWithProductInStock}
          selectedStore={selectedStore}
          handleClick={this.nearByClick}
          updateNearByList={this.updateNearByList}
        />
      </Wrapper>
    );
  }
}

const Wrapper = glamorous.div({
  display: 'flex',
  flexDirection: 'column',
  color: '#fff'
});

const MapWrapper = glamorous.div({
  width: '100%',
  height: '930px',
  position: 'relative'
});

const Title = glamorous.h1({
  fontSize: '48px',
  textAlign: 'center',
  lineHeight: '70px'
});

const SliderTitle = glamorous.p({
  fontSize: '16px',
  color: '#e4e4e4',
  fontFamily: 'LeroyMerlinSans Italic'
});

const ProductInfo = glamorous.h3({
  fontSize: '14px',
  textAlign: 'center',
  textTransform: 'capitalize',
  marginBottom: '66px'
});

const Circle = glamorous.div(({ height, width }) => ({
  boxSizing: 'content-box',
  border: '5px dotted #67cb33',
  transform: 'translate(-50%, -50%)',
  borderRadius: '50%',
  boxShadow: '0 0 0 9999999px rgba(0, 0, 0, 0.3)',
  pointerEvents: 'none',
  height,
  width
}));

const Radius = glamorous.div(({ left }) => ({
  fontSize: '16px',
  color: '#67cb33',
  margin: '0 auto 24px',
  textAlign: 'center',
  fontFamily: 'LeroyMerlinSans Light-Italic',
  position: 'absolute',
  left,
  top: 75
}));

const ControlsOverlay = glamorous.div({
  height: 30,
  position: 'absolute',
  bottom: 0,
  width: '100%'
});

const sliderStyle = { marginBottom: 24 };
const mapOptions = {
  gestureHandling: 'none',
  fullscreenControl: false,
  maxZoom: 17,
  minZoom: 8,
  zoomControl: false
};
