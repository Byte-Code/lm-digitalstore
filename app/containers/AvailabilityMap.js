import { connect } from 'react-redux';
import { compose, withState, withHandlers } from 'recompose';
import round from 'lodash/round';

import AvailabilityMap from '../components/AvailabilityMap';
import {
  getNearbyStoresWithStock,
  getStore,
  getNearbyStoresWithProductInStock,
  getSelectedNearbyStoreInfo
} from '../reducers/selectors';
import { requestFetchNearbyStores } from '../actions/storeActions';
import { trackStoreAvailabilityEvent } from '../actions/analyticsActions';

const mapStateToProps = (state, ownProps) => ({
  allNearbyStores: getNearbyStoresWithStock(state, ownProps),
  nearbyStoresWithProductInStock: getNearbyStoresWithProductInStock(state, ownProps),
  selectedStoreInfo: getSelectedNearbyStoreInfo(ownProps.selectedStore)(state, ownProps),
  homeStore: getStore(state)
});

const MapDispatchToProps = {
  requestFetchNearbyStores,
  trackStoreAvailabilityEvent
};

const mapKmZoom = {
  2: 13,
  12: 12,
  22: 11,
  32: 11,
  42: 10,
  50: 9
};

const enhance = compose(
  withState('radius', 'setRadius', 25),
  withState('zoom', 'setZoom', 11),
  withState('selectedStore', 'setSelectedStore', ''),
  withState('infoWindowOpen', 'setInfoWindow', false),
  withHandlers({
    selectStore: ({ setInfoWindow, setSelectedStore }) => value => {
      setSelectedStore(value, () => setInfoWindow(true));
    },
    closeInfoWindow: ({ setInfoWindow }) => () => setInfoWindow(false),
    handleChange: ({ setZoom }) => (e) => {
      setZoom(e.zoom);
    },
    handleSlide: ({ setRadius, setInfoWindow, setZoom }) => (e, v) => {
      const value = round(v);
      setZoom(mapKmZoom[value]);
      setInfoWindow(false, () => setRadius(v));
    }
  }),
  connect(mapStateToProps, MapDispatchToProps)
);

export default enhance(AvailabilityMap);
