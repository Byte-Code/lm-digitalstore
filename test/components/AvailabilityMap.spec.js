import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';

import AvailabilityMap from '../../app/components/AvailabilityMap';
import Marker from '../../app/components/Marker';
import InfoWindow from '../../app/components/InfoWindow';

// TODO find a way to test user interactions
const productCode = '123';
const productName = 'abc';
const allNearbyStores = fromJS([
  {
    code: '123',
    name: 'abc',
    latitude: 1231,
    longitude: 123213,
    address: {
      state: 'abc',
      street: 'bde',
      streetNumber: '1',
      zipCode: '213',
      city: 'milan'
    }
  },
  {
    code: '1234',
    name: 'asdsa',
    latitude: 12311,
    longitude: 123213,
    address: {
      state: 'asd',
      street: 'bae',
      streetNumber: '1',
      zipCode: '213',
      city: 'roma'
    }
  }
]);
const homeStore = fromJS({
  code: '123',
  name: 'current',
  gpsInformation: {
    x: 12311,
    y: 123213
  }
});
const requestFetchNearbyStores = jest.fn();

describe('AvailabilityMap', () => {
  const shallowWrapper = shallow(
    <AvailabilityMap
      productCode={productCode}
      productName={productName}
      allNearbyStores={allNearbyStores}
      homeStore={homeStore}
      requestFetchNearbyStores={requestFetchNearbyStores}
      nearbyStoresWithProductInStock={allNearbyStores}
      selectedStore=""
      radius={25}
      zoom={11}
      infoWindowOpen={false}
      selectStore={() => {}}
      handleChange={() => {}}
      handleSlide={() => {}}
      closeInfoWindow={() => {}}
    />
  );

  it('should initialize with the right interal state', () => {
    const expectedState = {
      minRadius: 2,
      maxRadius: 50,
      sliderWidth: 960
    };
    expect(shallowWrapper.state()).toEqual(expectedState);
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('should render a marker for each nearbyStore in range', () => {
    expect(shallowWrapper.find(Marker)).toHaveLength(2);
  });

  it('should not have rendered InfoWindow', () => {
    expect(shallowWrapper.find(InfoWindow)).toHaveLength(0);
  });
});
