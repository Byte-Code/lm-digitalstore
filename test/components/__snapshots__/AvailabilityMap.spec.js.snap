// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailabilityMap should initialize with the right interal state 1`] = `
<glamorous(div)>
  <glamorous(h1)>
    Verifica Disponibilit√†
  </glamorous(h1)>
  <glamorous(h3)>
    abc - REF. 123
  </glamorous(h3)>
  <glamorous.Div
    padding="0 20px 40px"
    position="relative"
  >
    <glamorous(p)>
      Seleziona il raggio di distanza dal negozio di current
    </glamorous(p)>
    <Slider
      axis="x"
      color="#67cb33"
      disableFocusRipple={false}
      disabled={false}
      max={50}
      min={2}
      onChange={[Function]}
      required={true}
      sliderStyle={
        Object {
          "marginBottom": 24,
        }
      }
      step={10}
      style={Object {}}
      value={25}
    />
    <glamorous(div)
      left={460}
    >
      25 km
    </glamorous(div)>
  </glamorous.Div>
  <glamorous(div)
    id="map"
  >
    <GoogleMap
      center={
        Object {
          "lat": 12311,
          "lng": 123213,
        }
      }
      debounced={true}
      distanceToMouse={[Function]}
      fullscreenControl={false}
      googleMapLoader={[Function]}
      hoverDistance={30}
      layerTypes={Array []}
      onChange={[Function]}
      options={
        Object {
          "fullscreenControl": false,
          "gestureHandling": "none",
          "maxZoom": 17,
          "minZoom": 8,
          "zoomControl": false,
        }
      }
      style={
        Object {
          "height": "100%",
          "margin": 0,
          "padding": 0,
          "position": "relative",
          "width": "100%",
        }
      }
      yesIWantToUseGoogleMapApiInternals={false}
      zoom={11}
    >
      <Marker
        code="123"
        handleClick={[Function]}
        isAvailable={false}
        isCurrentStore={true}
        lat={1231}
        lng={123213}
      />
      <Marker
        code="1234"
        handleClick={[Function]}
        isAvailable={false}
        isCurrentStore={false}
        lat={12311}
        lng={123213}
      />
      <glamorous(div)
        height={930}
        lat={12311}
        lng={123213}
        width={930}
      />
    </GoogleMap>
    <glamorous(div) />
  </glamorous(div)>
  <NearbyStores
    handleClick={[Function]}
    nearbyStores={
      Array [
        Object {
          "address": Object {
            "city": "milan",
            "state": "abc",
            "street": "bde",
            "streetNumber": "1",
            "zipCode": "213",
          },
          "code": "123",
          "latitude": 1231,
          "longitude": 123213,
          "name": "abc",
        },
        Object {
          "address": Object {
            "city": "roma",
            "state": "asd",
            "street": "bae",
            "streetNumber": "1",
            "zipCode": "213",
          },
          "code": "1234",
          "latitude": 12311,
          "longitude": 123213,
          "name": "asdsa",
        },
      ]
    }
    selectedStore=""
    slick={[Function]}
  />
</glamorous(div)>
`;
